[Kubernetes][Docs] GPU usage (#13325)
ray-project/ray: doc/source/cluster/kubernetes.rst

  kubectl delete -f ray/doc/kubernetes/ray-cluster.yaml

.. _k8s-gpus:

Using GPUs
----------

Using GPUs on Kubernetes requires some setup.

> Here's where to find the relevant documentation for `GKE`_, for `EKS`_, and for `AKS`_.

Reviewer: ```suggestion
Here's where to find the relevant documentation for GPU usage on `GKE`_, for `EKS`_, and for `AKS`_.
```

The `Ray Docker Hub <https://hub.docker.com/r/rayproject/>`_ hosts CUDA-based images packaged with Ray for use in Kubernetes pods. 
For example, the image ``rayproject/ray-ml:nightly-gpu`` is ideal for running GPU-based ML workloads with the most recent nightly build of Ray.
Read :ref:`here<docker-images>` for further details on Ray images. 

If you run into problems setting up GPUs for your Ray cluster on Kubernetes, please reach out to us at `<https://discuss.ray.io>`_. 

> 

Reviewer: Can you also mention that you need to explicitly change the resource requirements for the containers in your yaml?

Also, a snippet of relevant changes would be really helpful (which I think is only a couple lines of code)!

  kubectl delete -f ray/doc/kubernetes/ray-cluster.yaml

.. _k8s-gpus:

Using GPUs
----------

> Using GPUs on Kubernetes requires some setup.
> Here's where to find the relevant documentation for GPU usage on `GKE`_, for `EKS`_, and for `AKS`_.

Reviewer: ```suggestion
To use GPUs on Kubernetes, you will need to configure both your GKE setup and add additional values to your Ray cluster configuration.

For relevant documentation for GPU usage on different clouds, see instructions for `GKE`_, for `EKS`_, and for `AKS`_.
```

Author: Ah actually I guess it should say "Kubernetes setup" rather "GKE setup" (we should probably be cloud-agnostic)

Read :ref:`here<docker-images>` for further details on Ray images. 

Using Nvidia GPUs requires specifying the relevant resource `limits` in the container fields of your Kubernetes configurations.
(Kubernetes `sets <https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus/#using-device-plugins>`_ 
the GPU request equal to the limit.) The configuration for a pod running a Ray GPU image and
using one Nvidia GPU looks like this:

.. code-block:: yaml

  apiVersion: v1
  kind: Pod
  metadata:
   generateName: example-cluster-ray-worker
   spec:
    ...
    containers:
     - name: ray-node
       image: rayproject/ray:nightly-gpu
       ...
       resources:
        cpu: 1000m
        memory: 512Mi
       limits:
        memory: 512Mi
        nvidia.com/gpu: 1

GPU taints and tolerations
~~~~~~~~~~~~~~~~~~~~~~~~~~
.. note::

  Users using a managed Kubernetes service probably don't need to worry about this section. 

The `Nvidia gpu plugin`_ for Kubernetes applies `taints`_ to GPU nodes; these taints prevent non-GPU pods from being scheduled on GPU nodes.

> Managed Kubernetes services like GKE, EKS, and AKS automatically apply matching `tolerations`_ 
> to pods requesting GPU resources; tolerations are applied by means of Kubernetes's `ExtendedResourceToleration`_ `admission controller`_. 

Reviewer: ```suggestion
Managed Kubernetes services like GKE, EKS, and AKS automatically apply matching `tolerations`_ 
to pods requesting GPU resources. Tolerations are applied by means of Kubernetes's `ExtendedResourceToleration`_ `admission controller`_. 
```

If this admission controller is not enabled for your Kubernetes cluster, you may need to manually add a GPU toleration each of to your GPU pod configurations. For example,

.. code-block:: yaml

  apiVersion: v1
  kind: Pod
  metadata:
   generateName: example-cluster-ray-worker
   spec:
   ...
   tolerations:
   - effect: NoSchedule
     key: nvidia.com/gpu
     operator: Exists
   ...
   containers:
   - name: ray-node
     image: rayproject/ray:nightly-gpu
     ...

> Further reference and discussion
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
Further reference and discussion
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

> You can read about Kubernetes device plugins `here <https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/>`__,
> about Kubernetes GPU plugins `here <https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus>`__,
> and about Nvidia's GPU plugin for Kubernetes `here <https://github.com/NVIDIA/k8s-device-plugin>`__.

Reviewer: ```suggestion
Read about Kubernetes device plugins `here <https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/>`__,
about Kubernetes GPU plugins `here <https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus>`__,
and about Nvidia's GPU plugin for Kubernetes `here <https://github.com/NVIDIA/k8s-device-plugin>`__.
```
